solutions/cses-1076.mdx

```diff
@@ -87,3 +82,38 @@ int main() {
 	cout << endl;
 }
 ```
+
+## Solution 2: Order Statistic Tree / Fenwick Tree
+(Arpan Banerjee)
+
+We can directly use an [order statistic tree](https://codeforces.com/blog/entry/11080) (C++) to get the median while sliding the window across the array in $\mathcal{O}(n\log n)$ time.
+
+```cpp
+#include<bits/stdc++.h>
+#include<ext/pb_ds/assoc_container.hpp>
+#include<ext/pb_ds/tree_policy.hpp>
+using namespace std;
+using namespace __gnu_pbds;
+#define ordered_set tree<int, null_type,less_equal<int>, rb_tree_tag,tree_order_statistics_node_update>
+ 
+signed main(){
+        int n, k; cin >> n >> k; int a[n];
+        ordered_set oset;
+        for(int i = 0; i < n; i++){
+                int u; cin >> u; a[i]=u;
+                oset.insert(u);
+                if(i >= k) oset.erase(--oset.lower_bound(a[i-k]));
+                if(i >= k-1) cout << *oset.find_by_order((int)(k / 2.0 - 0.1)) << endl;
+        }
+}
+```
+
+Alternatively, if you do not use C++ (or if you do), you can use a [fenwick tree](https://usaco.guide/gold/PURS?lang=cpp#binary-indexed-tree) to simulate the aforementioned order statistic tree/indexed set:
+
+The fenwick array can be treated as a frequency array. For example, if `5` is inserted into the window, `fenwick_array[5]+=1`. This enables us to binary search for an integer $x$ such that 
+
+$$
+\sum_{i=1}^{x} \geq \lceil k/2 \rceil \text{ and } \sum_{i=1}^{x-1} \leq \lceil k/2 \rceil
+$$
+
+and find $k$ from there. Clearly, the values of the array would need to be compressed, as an array of size $10^9$ is infeasible. There will also be an extra $\log$ factor from using the fenwick tree `sum` operation, yielding a total time complexity of $\mathcal{O}(n\log^2{n})$. 
```

Comment:
you don't need an extra O(\log N) factor for the sum operation. you could first coordinate compress and update the values, then call the queries.

Reply:
we had convo ab this; it's fine